packages/core/rsc/render/render.tsx

```diff
@@ -0,0 +1,288 @@
+        tools:
+          tools == null
+            ? undefined
+            : Object.entries(tools).map(([name, tool]) => ({
+                type: 'function',
+                name,
+                description: tool.description,
+                parameters: convertZodToJSONSchema(tool.parameters),
+              })),
+      },
+      ...prepareCallSettings(settings),
+      inputFormat: validatedPrompt.type,
+      prompt: convertToLanguageModelPrompt(validatedPrompt),
+      abortSignal,
+    }),
+  );
+
+  const [stream, forkedStream] = result.stream.tee();
+
+  (async () => {
+    try {
+      // Consume the forked stream asynchonously.
+
+      let content = '';
+      let hasToolCall = false;
+
+      const reader = forkedStream.getReader();
+      while (true) {
+        const { done, value } = await reader.read();
+        if (done) break;
+
+        switch (value.type) {
+          case 'text-delta': {
+            content += value.textDelta;
+            handleRender(
+              [{ content, done: false, delta: value.textDelta }],
+              textRender,
+              ui,
+            );
+            break;
+          }
+
+          case 'tool-call-delta': {
+            hasToolCall = true;
+            break;
+          }
+
+          case 'tool-call': {
+            const toolName = value.toolName as keyof TOOLS & string;
+
+            if (!tools) {
+              throw new NoSuchToolError({ toolName: toolName });
+            }
+
+            const tool = tools[toolName];
+            if (!tool) {
+              throw new NoSuchToolError({
+                toolName,
+                availableTools: Object.keys(tools),
+              });
+            }
+
+            const parseResult = safeParseJSON({
+              text: value.args,
+              schema: tool.parameters,
+            });
+
+            if (parseResult.success === false) {
+              throw new InvalidToolArgumentsError({
+                toolName,
+                toolArgs: value.args,
+                cause: parseResult.error,
+              });
+            }
+
+            handleRender(
+              [
+                parseResult.value,
+                {
+                  toolName,
+                  toolCallId: value.toolCallId,
+                },
+              ],
+              tool.execute,
+              ui,
+            );
+
+            break;
+          }
+
+          case 'error': {
+            throw value.error;
+          }
+
+          case 'finish': {
+            // Nothing to do here.
+          }
+        }
+      }
+
+      if (hasToolCall) {
+        await finished;
+        ui.done();
+      } else {
+        handleRender([{ content, done: true }], textRender, ui);
+        await finished;
+        ui.done();
+      }
+    } catch (error) {
+      // During the stream rendering, we don't want to throw the error to the
+      // parent scope but only let the React's error boundary to catch it.
+      ui.error(error);
```

Comment:
Throwing the error and handling it here breaks the stream. Sometimes, the stream can be mostly good, only small parts fail (like a single tool call). An alternative could be a more granular error handling (catching/handling for each iteration inside the loop)

Reply:
Good point! Need to redesign that part a bit. Related: #1349.